lib/services/devices/deviceRegistryMemory.js

```diff
@@ -35,26 +35,41 @@ var registeredDevices = {},
  * @param {Object} newDevice           Device object to be stored
  */
 function storeDevice(newDevice, callback) {
-    if (registeredDevices[newDevice.id]) {
+    if (!registeredDevices[newDevice.service]) {
+        registeredDevices[newDevice.service] = {};
+    }
+
+    if (registeredDevices[newDevice.service][newDevice.id]) {
         callback(new errors.DuplicateDeviceId(newDevice.id));
     } else {
-        registeredDevices[newDevice.id] = newDevice;
-        registeredDevices[newDevice.id].creationDate = Date.now();
+        registeredDevices[newDevice.service][newDevice.id] = newDevice;
+        registeredDevices[newDevice.service][newDevice.id].creationDate = Date.now();
 
         logger.debug(context, 'Storing device with id [%s] and type [%s]', newDevice.id, newDevice.type);
         callback(null, newDevice);
     }
 }
 
 /**
- * Remove the device identified by its id and type.
+ * Remove the device identified by its id and service.
  *
- * @param {String} id           Device ID of the device to register.
+ * @param {String} id           Device ID of the device to remove.
+ * @param {String} service      Service of the device to remove.
+ * @param {String} subservice   Subservice inside the service for the removed device.
  */
-function removeDevice(id, callback) {
-    delete registeredDevices[id];
+function removeDevice(id, service, subservice, callback) {
+    var services = Object.keys(registeredDevices);
 
-    logger.debug(context, 'Removing device with id [%s].', id);
+    if (!subservice && !callback) {
+        callback = service;
+    }
+
+    for (var i = 0; i < services.length; i++) {
```

Comment:
This may cause unexpected behaviour in the case where the callback function is passed as the `service` parameter. The `length` property exists for functions and it indicates the number of arguments passed to the function. Consequently, it may only cause these unexpected behaviour in the case when a callback taking arguments is passed.

On the other hand, it seems that if no service is passed, the device is not deleted so maybe what is no longer needed is the code to control the arguments passed in case the `service` and `subservice` are required arguments.